Added Teradata driver

rawyler opened the pull request:
The driver has been used with a productive Teradata instance and works fine so far.
Tests were run as described in the wiki (ENGINES=h2,teradata lein test) using the Teradata Express image.
I managed to get down to 13 failures, 13 errors.
I sense that there isn't that much work left to reach 0 so I thought it's best to ask my remaining questions.

###### Questions:

#### 1.  Teradata: Duplicate definition of 'name' in NAMED phrase
    For example: breakout_test.clj:118 results in the following query:

```sql
SELECT
TOP 10 "test-data"."venues"."id" AS "id",
"test-data"."venues"."name" AS "name",
"test-data"."venues"."category_id" AS "category_id",
"test-data"."venues"."latitude" AS "latitude",
"test-data"."venues"."longitude" AS "longitude",
"test-data"."venues"."price" AS "price",
"categories__via__category_id"."name" AS "name"
FROM "test-data"."venues"
LEFT JOIN "test-data"."categories" "categories__via__category_id" ON "test-data"."venues"."category_id" = "categories__via__category_id"."id"
ORDER BY "categories__via__category_id"."name" DESC
```

For Teradata this results in "[Error 3863] [SQLState 42000] Duplicate definitio
n of 'name' in NAMED phrase."
I've seen that the implementations for Oracle and BigQuery support a de-duplication of aliases. Oracle does it if a limit is applied and BigQuery if an aggregation is made.

I wanted to ask if it is sufficent if the de-duplication is performed in apply-limit only.


#### 2. Database name: test_data does not exist
For example: nested_queries_test.clj:78 results in the following query:

```sql
SELECT "id", "longitude", "category_id", "price", "name", "latitude" FROM "test_data"."venues"
```

Somehow my Teradata implementation seems to generate wrong database identifiers when qualified using the identifier function of nested_queries_test.

Can you give me a hint?


#### 3. Date bucketing tests, some queries issued against wrong database

Some of the tests result in querying checkins:4-per-minute instead of checkins:4-per-hour or checkins:1-per-day. If i change the resulting query to use the correct database, the results are as expected by the test.

For example date_bucketing_test.clj:898 produces the following query:

```sql
SELECT count(*) AS "count" FROM "a-checkin-every-15-seconds"."checkins" WHERE trunc("a-checkin-every-15-seconds"."checkins"."timestamp", 'day') = trunc(CURRENT_TIMESTAMP, 'day')
```

although checkins:1-per-day should be used:

```sql
SELECT count(*) AS "count" FROM "a-checkin-every-86400-seconds"."checkins" WHERE trunc("a-checkin-every-86400-seconds"."checkins"."timestamp", 'day') = trunc(CURRENT_TIMESTAMP, 'day')
```

Do you have any idea why the tests in date_bucketing_test:885-... seem to use a-checkin-every-15-seconds although they request it differently?

#### 4. Order of the results in database_test.clj
Somehow the order of the returned result is not the same as the expected when running with ENGINES=h2,teradata. The data seems to be correct.

###### TODO (done)
-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

rawyler committed:
Organised imports. Removed mixin.

rawyler committed:
Added Teradata driver

salsakran commented:
@camsaul mind answering the questions that @rawyler brings up?

camsaul commented on src/metabase/driver/teradata.clj:
```diff
@@ -0,0 +1,306 @@
+     [helpers :as h]
+     types]
+    [metabase.models [field :as field]]
+    [clojure.tools.logging :as log]
```
We try to keep all our requires sorted. If you're using emacs you can have it do it for you automatically with the `cljr-clean-ns` command from the `clj-refactor` package (just make sure you `(setq cljr-favor-prefix-notation t)`). Otherwise take a look at other namespaces in the project for an example. We just sort them alphabetically, e.g. `clojure.tools.logging` should come before `metabase.models.field`.

camsaul commented:
> I wanted to ask if it is sufficent if the de-duplication is performed in apply-limit only.

That query you posted looks like it would still need alias deduplication even if you didn't apply a `TOP 10` limit. So you probably need to do it somewhere else besides `apply-limit` or you could still get into situations where a query with duplicate aliases could be generated.

camsaul commented on test/metabase/driver/teradata_test.clj:
```diff
@@ -0,0 +1,24 @@
+     [interface :refer [def-database-definition]]]
+    [toucan.db :as db]
+    [toucan.util.test :as tt])
+  (:import metabase.driver.teradata.TeradataDriver))
```
Did you just copy-paste all these required namespaces from somewhere else? It looks like you're using maybe two of them. If you're using Emacs you can just use `cljr-clean-ns` to clean up the ones that aren't used automatically

camsaul commented:
> Somehow my Teradata implementation seems to generate wrong database identifiers when qualified using the identifier function of nested_queries_test.

It looks to me like the venues table has `test_data` recorded as its schema so it's trying to qualify that as `schema`.`table-name`, which is why you get `test_data`.`venues`. Which way does Teradata expect it to be qualified? `database`.`schema`.`table-name`?

camsaul commented:
> Do you have any idea why the tests in date_bucketing_test:885-... seem to use a-checkin-every-15-seconds although they request it differently?

My best guess is there's something not working quite right in your test data code and it's causing it to reuse the same DB instead of creating new ones. So the unit tests think they're getting the checkin-every-day DB but it's fetching a different DB instead and thinks it has the right one. 

This might be an issue with reusing names. Different test DBs need to have different names or the test code will skip DB creation because it thinks the test DBs have already been created. Let me poke through the code and see if I can figure out what's going on

rawyler commented:
> Which way does Teradata expect it to be qualified? database.schema.table-name?

Teradata uses (user|database).table-name.column-name

camsaul commented:
So for tables it should be `database`.`table`, not `schema`.`table`?

camsaul commented:
> Order of the results in database_test.clj
> Somehow the order of the returned result is not the same as the expected when running with ENGINES=h2,teradata. The data seems to be correct.

Which test/line number is out-of-order? There's 40 or so tests in that file

camsaul commented on src/metabase/driver/teradata.clj:
```diff
@@ -0,0 +1,306 @@
+        (let [db-connection (sql/db->jdbc-connection-spec database)]
+          (run-query-without-timezone driver settings db-connection query))))))
+
+(def TeradataISQLDriverMixin
```
There's no reason to have a separate "mixin" here with a partial implementation of some of the driver methods. We only do that for the Postgres driver because the Redshift driver uses the same implementations. Since no other drivers are going to be using these Teradata implementations there's no need to make them available like this

rawyler replied:
ok

rawyler commented:
It would be `database.table` (the jdbc driver's `.getSchemas` method returns all the databases)

camsaul commented:
@rawyler is the only way to run Teradata locally with the Teradata Express VM image? That's fine for local development but I'm not sure how we're going to get that working on CI. Is there a Docker image?

Alternatively it sounds like it's available on the AWS Marketplace, but then we'd have to pay for that in order to test things

camsaul commented:
So `test_data` isn't the correct name of the database? In the query above it has `test-data`. Is that the right name for the DB? e.g. `"test-data"."venues"` instead of `"test_data"."venues"`?

rawyler commented:
Yes, `"test-data"."venues"` would be the correct one.

rawyler commented:
@camsaul The only thing I could find regarding the dockerization was this: https://community.teradata.com/t5/Database/Teradata-Express-for-Docker/m-p/71980#M1067
Instead of trying to get a Teradata Express running within a VirtualBox within a Docker container, I used the official VMWare image. I can look into this and check if a Teradata Express on VirtualBox within Docker would run.

rawyler commented:
As this Teradata Express within VirtualBox within Docker requires a lot of effort, I would like to run the tests against an AWS instance. Makes more sense and shouldn't be too expensive.

rawyler committed:
Merge branch 'master' into dev/teradata_driver

camsaul commented:
@rawyler what steps are needed to run this in AWS?

rawyler committed:
Merge branch 'master' into dev/teradata_driver

rawyler committed:
Added de-duplication for the Teradata driver

Using the same de-duplication function as the Oracle driver.

aedocw commented:
@camsaul there is a free instance of Teradata Database on AWS Marketplace (https://aws.amazon.com/marketplace/pp/B06Y4Z54R5?qid=1523485253250&sr=0-6&ref_=srh_res_product_title) which is launched with cloud formations.  Keeping the instance running full time would be unreasonably expensive, and I'm not sure if it would launch fast enough for the test runner?

Alternately, a low-performance solution is available - I got the Teradata Express VM image running in a VM under qemu.  It was pretty easy, and results in a DB running in a single instance (I think I used m2.large).  Would that work for testing if the version in marketplace doesn't work?

rawyler committed:
Merge branch 'master' into dev/teradata_driver

Commit:
Use supported docker image

jornh committed:
Update /bin/docker/Dockerfile to openjdk: as well

The `/bin/docker/Dockerfile` is the actual file used to build the actual released Docker images at https://hub.docker.com/r/metabase/metabase/.

The root folder `Dockerfile` is the one used for building https://hub.docker.com/r/metabase/metabase-head/ (that is, those we're built until the issue in https://github.com/metabase/metabase/issues/8444 appeared)

rawyler committed:
Merge branch 'master' into dev/teradata_driver

kdoh committed:
Dash card display fixes (#8777)

* fix scalar long title ellipsification and description

* fix table simple in night mode

* check for description

salsakran committed:
Merge pull request #8818 from metabase/master

Pull in snowflake + QP refactor + other stuff into 0.31

tlrobinson committed:
Fix minification

tlrobinson committed:
v0.31.0-RC1

tlrobinson committed:
Far-future cache header for JS/CSS files with cache-busting query strings

kdoh committed:
add scroll to outer container for database / schema selector (#8782)

add scroll to outer container for database / schema selector

camsaul committed:
Add Snowflake to list of supported databases in dox [ci skip]

camsaul committed:
Merge pull request #8852 from metabase/add-snowflake-to-databases-list

Add Snowflake to list of supported databases in dox

tlrobinson committed:
Merge pull request #8851 from metabase/fe-perf-fixes-0.31

Frontend perf fixes 0.31

kdoh committed:
scroll table selector (#8857)

tlrobinson committed:
Open links in new tab if command/'meta' key is pressed

camsaul committed:
Typo fixes and other code cleanup :shower: [ci bigquery]

camsaul committed:
Merge pull request #8865 from metabase/typo-fixes-and-other-code-cleanup

Typo fixes and other code cleanup :shower:

camsaul committed:
Remove BigQuery driver custom implementation of field->alias [ci bigquery]

camsaul committed:
Merge pull request #8867 from metabase/remove-bigquery-custom-alias-code

Remove BigQuery driver custom implementation of field->alias

camsaul committed:
BigQuery nested query support [ci drivers]

camsaul committed:
Merge pull request #8869 from metabase/big-nested-queries

BigQuery nested query support

camsaul committed:
Simplify SparkSQL driver; fix nested query support [ci drivers]

camsaul committed:
Fix BigQuery nested query support :wrench: [ci drivers]

camsaul committed:
Merge pull request #8870 from metabase/fix-sparksql-nested-query-support

Simplify SparkSQL driver; fix nested query support

kdoh committed:
Merge master fixes into release-0.31 (#8899)

* Default wrapper  padding left & right

For show cards padding on mobile view

* restructure visualization settings markup to fix label wrapping (#8816)

* Change query result hashing to convert integers to floats first

When the FE returns the query result metadata, any integer values will
be converted to a floating point number. Those doubles hash different
than the original integers which causes the hashes not to match and
the query to be reran when saving the question. This commit just
converts the integer to a double before computing the hash.

Fixes #8824

* Ensure queries to compute metadata include user information

When saving a question and the included metadata is incorrect or
different we will execute another query to compute that metadata. This
commit adds the `:info` map needed to include the user and query hash
information we use in the SQL remark.

Fixes #8825

* Metastore code SHOULD NOT barf it API response includes unrecognized keys

* Fix minification

* Update copyright year (#8711)

* Update copyright year

* 2018 year

* Could the admin nav bar use a little color? (#8819)

* purple admin nav bar

* decouple admin navbar color from accent7

* filter out hidden tables in table entity (#8831)

* Dropped tables should appear in a schema's data browse page

When the database had only a single schema, the dropped tables were
being hidden correctly. When multiple schemas were present, viewing
those tables their the schema route caused the inactive tables to
appear in the result.

Fixes #8822 fixes #8363

* Ensure consistent ordering of maps/sets when hashing metadata

This commit will ensure all hashes and sets are ordered before
serializing them to JSON and hashing that string. This will prevent
surprising behavior when the number of items in a map reach 9 or if
the order of elements changes between machines or JDK versions.

Fixes #8826

* Far-future cache header for JS/CSS files with cache-busting query strings

* Fix import on TableSettings page (#8829)

* fix import

* clean up nav

* Fix test checksum test failure on MySQL [ci drivers]

The newly added checksum tests fail when running on MySQL in
Circle. The MySQL tests include a secret key for encryption which are
used when computing the checksum. Because of that encryption, the
checksum string is different every time (it must be decrypted before
compared). The test isn't covering the encryption code but is only
concerned with ensure the datastructure hashes consitently. This
commit uses a `with-redefs` to set that secret key to `nil` so that
encryption won't be used on those two tests.

* Add better error messages to loading GeoJSON files

Previously the only error message was a schema validation failure. The
same error would be given for an unreachable host, unparsable JSON,
bad URL etc. This commit attempts to separate that out and provide a
better message in the failure case.

Fixes #8793

* Only include a CSV attachment for pulses when XLS is not selected

Previously we would always include a CSV attachment when there were
more rows/columns than would be included in the table. This ended up
being redundant for users that would prefer an XLS attachment and have
indicated that the pulse should include an XLS attachment. This commit
will only not attach a CSV if the user has select an XLS attachment
for that pulse.

Fixes #7302

* Add a C3P0 exclusion to Quartzite [ci drivers]

* Merge release 0.31.0 into master (#8858)

* Dash card display fixes (#8777)

* fix scalar long title ellipsification and description

* fix table simple in night mode

* check for description

* v0.31.0-RC1

* add scroll to outer container for database / schema selector (#8782)

add scroll to outer container for database / schema selector

* Fix minification

* Far-future cache header for JS/CSS files with cache-busting query strings

* Add Snowflake to list of supported databases in dox [ci skip]

* scroll table selector (#8857)

* Use compressed bytes for caching size comparison

Previously we would serialize rows as strings and use that as an
estimate for how big the cache would be. This commit changes that to
use actual compressed bytes to determine size.

* Switch to a stream based cache serialization/compression

This commit switches to a stream based nippy
compression/serialization. This has the benefit of being able to stop
serialization once the threshold for cache size has been reached.

Fixes #7479

* Bump the `java.classpath` and `tools.namespace` versions

Bumping `java.classpath` to `0.3.0` fixes the `classpath` function on
JDK 9+.

* Don't strip insights from response in public links

* update tests

* Bump Jetty server version

* Better-organized HTTP client used for tests and add full-response option.

* Use middleware impl for test util fn bind-current-user

* Add some i18n tags and wrap in str where missing

* Namespace form & indentation cleanup; docstring improvments

* Remove dead code in tests :skull:

* MBQL schema tweaks; make SQL QP nested-query-level available to public

* Ignore results_metadata in revision tests

* Add test for updating permissions graph from API :white_check_mark:

* Add u/hexadecimal-string? util fn

* Add query-perms/can-run-query? helper fn :closed_lock_with_key:

* Record database_id in QueryExecution; record query in Query :page_with_curl:

* Add a few util fns; other minor code tweaks

* Change query perms-set fn to take kwargs for opts :white_check_mark:

* Support backslashes in DB/schema/table names for permissions checks

* Bump docstring checker -> 1.0.3

* Fix bug where Snowflake driver used wrong DB name to filter tables [ci snowflake]

* Add new bird-count dataset; fix division by zero in MBQL expressions

* Add test profiling info & don't suppress dataset load output [ci drivers]

* Remove unused default-schema; share Presto catalog for all test data [ci drivers]

* Add format-name for SparkSQL test extensions [ci drivers]

* Fix presto tests :wrench: [ci presto]

* handle smaller smart scalar cards better (#8897)

* handle smaller cards better

* handle automatic dashboard sidebar case

* make more specific

* Fix presto test [ci presto] (#8900)

kdoh committed:
test-run

tlrobinson committed:
Update translations

camsaul committed:
v0.31.0

iethree committed:
Fix unspecified PG database connection strings

This issue can occur if the `MB_DB_DBNAME` variable has not been
specified. The `:dbname` key has a nil value which causes Ring's
`form-encode` function to fail. This commit explicitly removes the
`:dbname` as it's extra anyway and will also remove map entries that
have a nil key or value before encoding it.

senior committed:
Merge pull request #8913 from metabase/fix-omitted-pg-dbname-issue

Fix unspecified PG database connection strings

camsaul committed:
Merge pull request #8459 from AndreaGiardini/dockerImageSupp

Use supported java docker image

camsaul committed:
Fix migration failures :wrench: [ci drivers]

camsaul committed:
Remove some unused files from uberjar

camsaul committed:
Handle missing columns in MongoDB results :male-detective: [ci mongo]

camsaul committed:
Merge pull request #8944 from metabase/tree-shaking

Shave 10-20MB off uberjar size

camsaul committed:
Merge pull request #8952 from metabase/handle-missing-columns-in-mongodb-results

Handle missing columns in MongoDB results :male-detective: [ci mongo]

camsaul committed:
Auto-bucket datetime field in filter clauses :calendar: [ci drivers]

cgopalan committed:
Fix QP include-current behavior for previous/next time intervals :timer_clock:

camsaul committed:
Merge pull request #8942 from metabase/fix-migration-failures

Fix migration failures :wrench: [ci drivers]

MarkRx committed:
Add ability to copy a dashboard :page_facing_up:

camsaul committed:
Revert tree-shaking

camsaul committed:
Merge pull request #8976 from metabase/revert-tree-shaking

Revert tree-shaking

camsaul committed:
Merge pull request #8974 from metabase/paychex-fix-custom-ui-include-today

Fix QP include-current behavior for previous/next time intervals

camsaul committed:
Merge pull request #8975 from metabase/paychex-copy-dashboard

Add ability to copy a dashboard

camsaul committed:
Merge pull request #8973 from metabase/fix-8927

Auto-bucket datetime fields in filter clauses

camsaul committed:
Gracefully handle normalization of invalid queries :warning:

camsaul committed:
Merge pull request #8979 from metabase/gracefully-handle-normalization-failures

Gracefully handle normalization of invalid queries

camsaul committed:
Fail gracefully when migrating invalid BigQuery legacy SQL questions :skull:

camsaul committed:
Merge pull request #8980 from metabase/fail-gracefully-when-migrating-invalid-bigquery-cards

Fail gracefully when migrating invalid BigQuery legacy SQL questions

camsaul committed:
Fix nested queries with JOINs :hatching_chick: [ci drivers]

rawyler committed:
disable nested queries for teradata

camsaul committed:
Merge pull request #8981 from metabase/fix-nested-queries-with-joins

Fix nested queries with JOINs

camsaul committed:
Fix Mongo & Druid multiple ags of same type :v: [ci drivers]

camsaul committed:
Add expression_name to QP cols results

camsaul committed:
Merge pull request #8987 from metabase/fix-mongo-druid-multiple-ags-same-type

Fix Mongo & Druid multiple ags of same type

camsaul committed:
Fix non-admins w/ perms not able to create new collection in root :unlock:

camsaul committed:
Merge pull request #8990 from metabase/allow-non-admins-to-create-children-of-root-collection

Fix non-admins w/ perms not able to create new collection in root

camsaul committed:
Fix migration failures due to timeouts :stopwatch:

camsaul committed:
Merge pull request #8991 from metabase/fix-migration-timeouts

Fix migration failures due to timeouts :stopwatch:

camsaul committed:
Test fix :wrench:

tlrobinson committed:
Fix column actions for custom expressions

tlrobinson committed:
Fix crash on custom aggregations

camsaul committed:
Merge pull request #8989 from metabase/add-expression_name-to-QP-cols

Add expression_name back to QP cols results

camsaul committed:
v0.31.1

tlrobinson committed:
Remove stray paren/bracket

tsadauskas committed:
Added missing run_metabase.sh script from the builder sources.

tlrobinson committed:
Fix currency formatting of negative numbers when currency_in_header = true

camsaul committed:
Fix handling dimension parameters using a foreign key :key:
[ci drivers]

camsaul committed:
Auto-bucket type/DateTime field literals where appropriate :calendar:
[ci drivers]

camsaul committed:
Update MongoDB driver :up:
[ci mongo]

camsaul committed:
Bump version -> 0.31.2 :up:

camsaul committed:
Merge pull request #9030 from metabase/bump-mongoDB-dependency

Update MongoDB driver

camsaul committed:
QB should use normalized cumulative count/sum MBQL names :abcd:

camsaul committed:
Cache Slack files-channel to avoid rate limiting :floppy_disk:

camsaul committed:
Merge pull request #9031 from metabase/fix-8979

QB should use normalized cumulative count/sum MBQL names :abcd:

camsaul committed:
Merge pull request #9032 from metabase/add-ttl-caching-to-slack-files-channel

Cache Slack files-channel to avoid rate limiting :floppy_disk:

camsaul committed:
Merge pull request #9025 from metabase/fix-9017

Fix handling dimension parameters using a foreign key

camsaul committed:
Merge pull request #9029 from metabase/fix-9007

Auto-bucket type/DateTime field literals where appropriate :calendar:

camsaul committed:
Don't log info messages or async load Liquibase during compilation :construction_worker:

camsaul committed:
Warn on Postgres conn str w/ ssl=true but not sslmode=require :warning:

tlrobinson committed:
Merge pull request #8860 from metabase/command-click-new-tab

Open links in new tab if command/'meta' key is pressed

tlrobinson committed:
Disable tests that are failing on CI

tlrobinson committed:
Merge pull request #9053 from metabase/disable-broken-tests

Disable tests that are failing on CI

tlrobinson committed:
Merge branch 'release-0.31.2' of github.com:metabase/metabase into fix-negative-currency-formatting

tlrobinson committed:
Merge pull request #9022 from metabase/fix-negative-currency-formatting

Fix currency formatting of negative numbers when currency_in_header = true

kdoh committed:
check can_write when displaying dashboard header buttons (#9047)

* check can_write when displaying header buttons

* check can_write when displaying header buttons

* don't hide duplicate

camsaul committed:
Merge pull request #9043 from metabase/extra-logging-for-DynamicClassLoadHelper-errors

Don't log info messages or async load Liquibase during compilation

camsaul committed:
Merge pull request #9045 from metabase/add-warning-message-when-connecting-to-pg-without-sslmode-require

Warn on Postgres conn str w/ ssl=true but not sslmode=require :warning:

kdoh committed:
Fix popover scrolling (#9046)

* ensure group and accordian lists scroll when the window is too small

* CR fixes

mazameli committed:
update languages and add german

tlrobinson committed:
Fix date formatting settings for SQL queries

mazameli committed:
Merge pull request #9066 from metabase/update-strings

Update released languages' strings and add German (yay!)

tlrobinson committed:
Merge pull request #9068 from metabase/sql-date-formatting

Fix date formatting settings for SQL queries

kdoh committed:
remove unnecessary height declaration (#9076)

rawyler committed:
Merge remote-tracking branch 'origin-metabase/release-0.31.2' into dev/teradata_driver

tlrobinson commented:
Hey all,

Thank you for this PR and apologies for the long delay responding. We've recently done some work to support drivers as plugins in Metabase and as of the next release (v0.32) we will be asking driver developers to first publish new drivers in their own repositories (typically named `metabase-*-driver`). If the driver fully passes the driver tests we will link to it from our documentation, and if a driver gains significant usage we will consider merging it into Metabase itself.

Here is some documentation about publishing a driver as a plugin in a separate repository: https://github.com/metabase/metabase/wiki/Writing-a-Driver:-Packaging-a-Driver-&-Metabase-Plugin-Basics#drivers-shipped-as-3rd-party-plugins

Here are two example drivers: https://github.com/metabase/sudoku-driver https://github.com/metabase/crate-driver

We still have some work to do on the documentation. You can follow progress and provide feedback on that here: https://github.com/metabase/metabase/issues/9348

Thank you.

rawyler commented:
Thanks for your feedback.
I will perform the transition to a separate repository soon.

funsign commented:
Hi,
I’m very interested by this teradata driver. When he will be merge?

Thx

rawyler commented:
The Teradata driver has been migrated to the plugin `metabase-teradata-driver` https://github.com/swisscom-bigdata/metabase-teradata-driver

funsign commented:
> The Teradata driver has been migrated to the plugin `metabase-teradata-driver` https://github.com/swisscom-bigdata/metabase-teradata-driver

Hi,
thank you very much, tested and approved !! 

Bye
Antoine

arita37 commented:
Very good thanks.
